File: src/main/java/io/github/qingmo/json/JSONObject.java
/*
 * MIT License
 *
 * Copyright (c) 2021 qingmo(eagleqingluo@gmail.com)
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */
package io.github.qingmo.json;

import io.github.qingmo.json.internal.TypeUtils;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class JSONObject implements Map<String, Object> {

    private Map<String, Object> map;

    public JSONObject(Map<String, Object> map) {
        this.map = map;
    }

    public JSONObject() {
        this.map = new HashMap<>();
    }

    public JSONObject(Integer size) {
        this.map = new HashMap<>(size);
    }

    public String toJSONString() {
        return JSON.toJSONString(this.map);
    }

    public JSONObject getJSONObject(String key) {
        Object value = map.get(key);
        if (value == null) {
            return null;
        }
        if (value instanceof JSONObject) {
            return (JSONObject) value;
        }
        return JSON.parseObject(JSON.toJSONString(value));
    }

    public JSONArray getJSONArray(String key) {
        Object value = map.get(key);
        if (value == null) {
            return new JSONArray();
        }
        if (value instanceof JSONArray) {
            return (JSONArray) value;
        }
        return JSON.parseArray(JSON.toJSONString(value));
    }


    public <T> T getObject(String key, Class<T> clazz) {
        Object obj = map.get(key);
        if (obj == null) {
            return null;
        }
        return JSON.parseObject(JSON.toJSONString(obj), clazz);
    }

    public Boolean getBoolean(String key) {
        Object value = map.get(key);
        if (value == null) {
            return null;
        }
        return TypeUtils.castToBoolean(value);
    }

    public Boolean getBooleanValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToBoolean(value);
        } catch (Exception ignore) {
            return false;
        }
    }

    public Byte getByte(String key) {
        Object value = map.get(key);
        return TypeUtils.castToByte(value);
    }

    public Byte getByteValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToByte(value);
        } catch (Exception ignore) {
            return (byte) 0;
        }
    }

    public Short getShort(String key) {
        Object value = map.get(key);
        return TypeUtils.castToShort(value);
    }

    public Short getShortValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToShort(value);
        } catch (Exception ignore) {
            return (short) 0;
        }
    }

    public Integer getInteger(String key) {
        Object value = map.get(key);
        return TypeUtils.castToInt(value);
    }

    public Integer getIntValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToInt(value);
        } catch (Exception ignore) {
            return 0;
        }
    }

    public Long getLong(String key) {
        Object value = map.get(key);
        return TypeUtils.castToLong(value);
    }

    public Long getLongValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToLong(value);
        } catch (Exception ignore) {
            return 0L;
        }
    }

    public Float getFloat(String key) {
        Object value = map.get(key);
        return TypeUtils.castToFloat(value);
    }

    public Float getFloatValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToFloat(value);
        } catch (Exception ignore) {
            return 0f;
        }
    }

    public Double getDouble(String key) {
        Object value = map.get(key);
        return TypeUtils.castToDouble(value);
    }

    public Double getDoubleValue(String key) {
        Object value = map.get(key);
        try {
            return TypeUtils.castToDouble(value);
        } catch (Exception ignore) {
            return 0.0;
        }
    }

    public BigDecimal getBigDecimal(String key) {
        Object value = map.get(key);
        return TypeUtils.castToBigDecimal(value);
    }

    public BigInteger getBigInteger(String key) {
        Object value = map.get(key);
        return TypeUtils.castToBigInteger(value);
    }

    public String getString(String key) {
        Object value = map.get(key);
        if (value == null) {
            return null;
        }
        return value.toString();
    }

    public <T> T toJavaObject(Class<T> clazz) {
        if (clazz == Map.class || clazz == JSONObject.class) {
            return (T) this;
        } else {
            return JSON.parseObject(toJSONString(), clazz);
        }
    }

    @Override
    public String toString() {
        return toJSONString();
    }


    @Override
    public Object remove(Object key) {
        return map.remove(key.toString());
    }


    @Override
    public int size() {
        return map.size();
    }

    @Override
    public boolean isEmpty() {

        return map.isEmpty();
    }

    @Override
    public boolean containsKey(Object key) {
        return map.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return map.containsValue(value);
    }

    @Override
    public Object get(Object key) {
        return map.get(key);
    }

    @Override
    public Object put(String key, Object value) {
        return map.put(key, value);
    }


    @Override
    public void putAll(Map<? extends String, ?> m) {
        map.putAll(m);
    }

    @Override
    public void clear() {
        map.clear();
    }

    @Override
    public Set<String> keySet() {
        return map.keySet();
    }

    @Override
    public Collection<Object> values() {
        return map.values();
    }

    @Override
    public Set<Entry<String, Object>> entrySet() {
        return map.entrySet();
    }

    @Override
    public boolean equals(Object o) {
        return map.equals(o);
    }

    @Override
    public int hashCode() {
        return map.hashCode();
    }
}

File: src/main/kotlin/io/github/qingmo/json/JSONObject.kt
/*
 * Copyright (c) 2020-2021, chaos (eagleqingluo@gmail.com).
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.github.qingmo.json

import io.github.qingmo.json.JSON.parseObject
import java.math.BigDecimal
import java.math.BigInteger
import java.util.HashMap

class JSONObject : MutableMap<Any, Any?> {
    private var map: MutableMap<Any, Any?>

    constructor(map: MutableMap<Any, Any?>) {
        this.map = map
    }

    constructor() {
        map = HashMap()
    }

    constructor(size: Int) {
        map = HashMap(size)
    }

    fun toJSONString(): String {
        return JSON.toJSONString(map)
    }

    fun getJSONObject(key: String): JSONObject? {
        val value = map.get(key) ?: return null
        if (value is JSONObject) {
            return value
        }
        return JSON.parseObject(JSON.toJSONString(value))
    }

    fun getJSONArray(key: String): JSONArray {
        val value = map[key] ?: return JSONArray()
        if (value is JSONArray) {
            return value
        }
        return JSON.parseArray(JSON.toJSONString(value))
    }

    
    fun <T> getObject(key: String, clazz: Class<T>): T? {
        val obj = map[key] ?: return null
        return parseObject(JSON.toJSONString(obj), clazz)
    }

    fun getBoolean(key: String): Boolean? {
        val value = get(key) ?: return null
        return TypeUtils.castToBoolean(value)
    }

    fun getBooleanValue(key: String): Boolean {
        val value = get(key)
        val booleanVal: Boolean = TypeUtils.castToBoolean(value) ?: return false
        return booleanVal
    }

    fun getByte(key: String): Byte? {
        val value = get(key)
        return TypeUtils.castToByte(value)
    }

    fun getByteValue(key: String): Byte {
        val value = get(key)
        return TypeUtils.castToByte(value) ?: return 0
    }

    fun getShort(key: String): Short? {
        val value = get(key)
        return TypeUtils.castToShort(value)
    }

    fun getShortValue(key: String): Short {
        val value = get(key)
        return TypeUtils.castToShort(value) ?: return 0
    }

    fun getInteger(key: String): Int? {
        val value = get(key)
        return TypeUtils.castToInt(value)
    }

    fun getIntValue(key: String): Int {
        val value = get(key)
        return TypeUtils.castToInt(value) ?: return 0
    }

    fun getLong(key: String): Long? {
        val value = get(key)
        return TypeUtils.castToLong(value)
    }

    fun getLongValue(key: String): Long {
        val value = get(key)
        return TypeUtils.castToLong(value) ?: return 0L
    }

    fun getFloat(key: String): Float? {
        val value = get(key)
        return TypeUtils.castToFloat(value)
    }

    fun getFloatValue(key: String): Float {
        val value = get(key)
        return TypeUtils.castToFloat(value) ?: return 0f
    }

    fun getDouble(key: String): Double? {
        val value = get(key)
        return TypeUtils.castToDouble(value)
    }

    fun getDoubleValue(key: String): Double {
        val value = get(key)
        return TypeUtils.castToDouble(value) ?: return 0.0
    }

    fun getBigDecimal(key: String): BigDecimal? {
        val value = get(key)
        return TypeUtils.castToBigDecimal(value)
    }

    fun getBigInteger(key: String): BigInteger? {
        val value = get(key)
        return TypeUtils.castToBigInteger(value)
    }

    fun getString(key: String): String? {
        val value = get(key) ?: return null
        return value.toString()
    }

    fun <T> toJavaObject(clazz: Class<T>): T {
        return if (clazz == MutableMap::class.java || clazz == JSONObject::class.java || clazz == JSONArray::class.java) {
            @Suppress("UNCHECKED_CAST")
            this as T
        } else parseObject(toJSONString(), clazz)
    }

    override fun toString(): String {
        return toJSONString()
    }

    override fun remove(key: Any): Any? {
        return map.remove(key)
    }

    override val size: Int
        get() = map.size

    override fun isEmpty(): Boolean {
        return map.isEmpty()
    }

    override fun containsKey(key: Any): Boolean {
        return map.containsKey(key)
    }

    override fun containsValue(value: Any?): Boolean {
        return map.containsValue(value)
    }

    override fun get(key: Any): Any? {
        return map.get(key)
    }

    override fun put(key: Any, value: Any?): Any? {
        return map.put(key, value)
    }

    override fun putAll(from: Map<out Any, Any?>) {
        map.putAll(from)
    }

    override fun clear() {
        map.clear()
    }

    override val keys: MutableSet<Any>
        get() = map.keys

    override val values: MutableCollection<Any?>
        get() = map.values

    override val entries: MutableSet<MutableMap.MutableEntry<Any, Any?>>
        get() = map.entries

    override fun equals(other: Any?): Boolean {
        return map == other
    }

    override fun hashCode(): Int {
        return map.hashCode()
    }
}

